### PR TITLE
turn pcre2 dependency into dlopen() dependency

### PR DESCRIPTION
**Pull request by poettering**
Let's make use of pcre2 if it's available, but not strictly pull it in.

### CODE REVIEW

File: src/journal/pcre2-dlopen.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                        &sym_pcre2_get_error_message, "pcre2_get_error_message_8",
+                        &sym_pcre2_match, "pcre2_match_8",
+                        &sym_pcre2_get_ovector_pointer, "pcre2_get_ovector_pointer_8",
+                        NULL);
+        if (r < 0)
+                return r;
+
+        /* Note that we never release the reference here, because there's no realy reason to, after all this
+         * was traditionally a regular shared library dependency which lives forever too. */
+        pcre2_dl = TAKE_PTR(pcre2_dl);
```
Comment (flo-wer) [inline src/journal/pcre2-dlopen.c]:
```suggestion
        pcre2_dl = TAKE_PTR(dl);
```

Comment (poettering) [reply src/journal/pcre2-dlopen.c]: ouch! indeed!

**Comment by mrc0mmand** [comment]
The CentOS CI fail is unrelated, see #16105...

File: src/journal/pcre2-dlopen.c [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                        &sym_pcre2_code_free, "pcre2_code_free_8",
+                        &sym_pcre2_compile, "pcre2_compile_8",
+                        &sym_pcre2_get_error_message, "pcre2_get_error_message_8",
+                        &sym_pcre2_match, "pcre2_match_8",
+                        &sym_pcre2_get_ovector_pointer, "pcre2_get_ovector_pointer_8",
+                        NULL);
+        if (r < 0)
+                return r;
+
+        /* Note that we never release the reference here, because there's no realy reason to, after all this
```
Comment (clintonroy) [inline src/journal/pcre2-dlopen.c]: realy -> really

**Comment by poettering** [comment]
Force pushed new version, only chance is the typo fix @clintonroy found.

File: src/basic/dlfcn-util.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                                           "Can't find symbol %s: %s", symbol, dlerror());
+                        va_end(ap);
+                        return r;
+                }
+
+                *fn = tfn;
+        }
+
+        va_end(ap);
+        return 0;
```
Comment (keszybz) [inline src/basic/dlfcn-util.c:39]: I hope it's the same as the one in the other PR ;)

File: src/journal/pcre2-dlopen.c [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        _cleanup_(dlclosep) void *dl = NULL;
+        int r;
+
+        if (pcre2_dl)
+                return 0; /* Already loaded */
+
+        dl = dlopen("libpcre2-8.so.0", RTLD_LAZY);
+        if (!dl)
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "PCRE2 support is not installed: %s", dlerror());
```
Comment (keszybz) [inline src/journal/pcre2-dlopen.c:28]: Hmm, this should probably mention the name we tried to dlopen...

File: src/journal/pcre2-dlopen.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                        dl,
+                        LOG_ERR,
+                        &sym_pcre2_match_data_create, "pcre2_match_data_create_8",
+                        &sym_pcre2_match_data_free, "pcre2_match_data_free_8",
+                        &sym_pcre2_code_free, "pcre2_code_free_8",
+                        &sym_pcre2_compile, "pcre2_compile_8",
+                        &sym_pcre2_get_error_message, "pcre2_get_error_message_8",
+                        &sym_pcre2_match, "pcre2_match_8",
+                        &sym_pcre2_get_ovector_pointer, "pcre2_get_ovector_pointer_8",
+                        NULL);
```
Comment (keszybz) [inline src/journal/pcre2-dlopen.c:40]: This is slowly crossing over into the "somewhat ugly" territory. I guess it's still OK, but I hope we can figure out something better in the long term.

**Pull request merged**